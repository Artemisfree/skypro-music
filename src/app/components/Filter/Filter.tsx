'use client'

import React, { useState, useEffect, MouseEvent } from 'react'
import { getAllTracks } from '@/app/api'
import styles from './Filter.module.css'

interface Track {
	author: string
	genre: string
	release_date: string
}

interface Position {
	top: number
	left: number
}

const Filter: React.FC = () => {
	const [tracks, setTracks] = useState<Track[]>([])
	const [uniqueAuthors, setUniqueAuthors] = useState<string[]>([])
	const [uniqueGenres, setUniqueGenres] = useState<string[]>([])
	const [uniqueYears, setUniqueYears] = useState<number[]>([])
	const [activeFilter, setActiveFilter] = useState<string | null>(null)
	const [dropdownPosition, setDropdownPosition] = useState<Position>({
		top: 0,
		left: 0,
	})

	useEffect(() => {
		const fetchTracks = async () => {
			try {
				const data = await getAllTracks()
				setTracks(data)
				extractUniqueValues(data)
			} catch (error) {
				console.error('Ошибка при получении треков:', error)
			}
		}

		fetchTracks()
	}, [])

	const extractUniqueValues = (tracks: Track[]) => {
		const authors = new Set<string>()
		const genres = new Set<string>()
		const years = new Set<number>()

		tracks.forEach(track => {
			authors.add(track.author)
			genres.add(track.genre)
			const year = new Date(track.release_date).getFullYear()
			years.add(year)
		})

		setUniqueAuthors(Array.from(authors))
		setUniqueGenres(Array.from(genres))
		setUniqueYears(Array.from(years).sort((a, b) => a - b))
	}

	const handleFilterClick = (
		filter: string,
		event: MouseEvent<HTMLDivElement>
	) => {
		const button = event.currentTarget
		const rect = button.getBoundingClientRect()
		setDropdownPosition({
			top: rect.bottom + window.scrollY + 10,
			left: rect.left + window.scrollX,
		})
		setActiveFilter(prevFilter => (prevFilter === filter ? null : filter))
	}

	return (
		<div className={styles.centerblock__filter}>
			<div className={styles.filter__title}>Искать по:</div>
			<div
				className={`${styles.filter__button} ${
					activeFilter === 'author' ? styles.active : ''
				}`}
				onClick={event => handleFilterClick('author', event)}
			>
				исполнителю
			</div>
			{activeFilter === 'author' && (
				<div
					className={`${styles.filter__dropdown} ${styles.dropdown_performer} ${styles.active}`}
					style={{ top: dropdownPosition.top, left: dropdownPosition.left }}
				>
					{uniqueAuthors.map(author => (
						<div key={author} className={styles.filter__item}>
							{author}
						</div>
					))}
				</div>
			)}
			<div
				className={`${styles.filter__button} ${
					activeFilter === 'year' ? styles.active : ''
				}`}
				onClick={event => handleFilterClick('year', event)}
			>
				году выпуска
			</div>
			{activeFilter === 'year' && (
				<div
					className={`${styles.filter__dropdown} ${styles.dropdown_year} ${styles.active}`}
					style={{ top: dropdownPosition.top, left: dropdownPosition.left }}
				>
					{uniqueYears.map(year => (
						<div key={year} className={styles.filter__item}>
							{year}
						</div>
					))}
				</div>
			)}
			<div
				className={`${styles.filter__button} ${
					activeFilter === 'genre' ? styles.active : ''
				}`}
				onClick={event => handleFilterClick('genre', event)}
			>
				жанру
			</div>
			{activeFilter === 'genre' && (
				<div
					className={`${styles.filter__dropdown} ${styles.dropdown_genre} ${styles.active}`}
					style={{ top: dropdownPosition.top, left: dropdownPosition.left }}
				>
					{uniqueGenres.map(genre => (
						<div key={genre} className={styles.filter__item}>
							{genre}
						</div>
					))}
				</div>
			)}
		</div>
	)
}

export default Filter
