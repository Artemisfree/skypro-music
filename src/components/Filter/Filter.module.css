/* .centerblock__filter {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 51px;
}

.filter__title {
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 24px;
  margin-right: 15px;
}

.filter__button {
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 24px;
  border: 1px solid #ffffff;
  border-radius: 60px;
  padding: 6px 20px;
  cursor: pointer;
}

.filter__button:not(:last-child) {
  margin-right: 10px;
}

.filter__dropdown {
  background-color: #313131;
  border-radius: 8px;
  padding: 10px;
  max-height: 200px;
  overflow-y: auto;
  position: absolute;
  top: 50px;
  left: 0;
  z-index: 10;
  display: none;
}

.filter__button.active .filter__dropdown {
  display: block;
}

.dropdown_performer,
.dropdown_year,
.dropdown_genre {
  width: 200px;
}

.filter__item {
  font-size: 14px;
  color: #ffffff;
  padding: 5px 0;
  cursor: pointer;
}

.filter__item:hover {
  color: #B672FF;
  text-decoration: underline;
}

.filter__button:hover,
.filter__button:active,
.filter__button.active {
  border-color: #B672FF;
  color: #B672FF;
}

.filter__dropdown.active {
  display: block;
} */


.centerblock__filter {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-bottom: 51px;
}

.filter__title {
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 24px;
  margin-right: 15px;
}

.filter__button {
  position: relative;
  font-style: normal;
  font-weight: 400;
  font-size: 16px;
  line-height: 24px;
  border: 1px solid #ffffff;
  border-radius: 60px;
  padding: 6px 20px;
  cursor: pointer;
}

.filter__button:not(:last-child) {
  margin-right: 10px;
}

.filter__dropdown {
  background-color: #313131;
  border-radius: 8px;
  padding: 10px;
  max-height: 200px;
  overflow-y: auto;
  position: absolute;
  top: 50px;
  left: 0;
  z-index: 10;
  display: none;
}

.filter__button.active .filter__dropdown {
  display: block;
}

.dropdown_performer,
.dropdown_year,
.dropdown_genre {
  width: 200px;
}

.filter__item {
  font-size: 14px;
  color: #ffffff;
  padding: 5px 0;
  cursor: pointer;
}

.filter__item:hover {
  color: #B672FF;
  text-decoration: underline;
}

.filter__item.active,
.filter__button:hover,
.filter__button:active,
.filter__button.active {
  border-color: #B672FF;
  color: #B672FF;
}

.filter__dropdown.active {
  display: block;
}

.filter__badge {
    position: absolute;
    top: -10px;
    right: -10px;
    width: 26px;
    height: 25.5px;
    background-color: #B672FF;
    border-radius: 50%;
    display: flex;
    align-items: center;
    justify-content: center;
    color: white;
    font-size: 14px;
    font-weight: bold;
}