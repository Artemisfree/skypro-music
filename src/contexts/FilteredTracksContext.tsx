'use client'

import React, { createContext, useState, useContext } from 'react'

interface FilteredTracksContextType {
    activeGenres: string[]
    setActiveGenres: React.Dispatch<React.SetStateAction<string[]>>
    activeAuthors: string[]
    setActiveAuthors: React.Dispatch<React.SetStateAction<string[]>>
    searchKeyword: string
    setSearchKeyword: React.Dispatch<React.SetStateAction<string>>
    sortOrder: string
    setSortOrder: React.Dispatch<React.SetStateAction<string>>
}

const FilteredTracksContext = createContext<FilteredTracksContextType | undefined>(undefined)

export const FilteredTracksProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    const [activeGenres, setActiveGenres] = useState<string[]>([])
    const [activeAuthors, setActiveAuthors] = useState<string[]>([])
    const [searchKeyword, setSearchKeyword] = useState<string>('')
    const [sortOrder, setSortOrder] = useState<string>('default')

    return (
        <FilteredTracksContext.Provider value={{ activeGenres, setActiveGenres, activeAuthors, setActiveAuthors, searchKeyword, setSearchKeyword, sortOrder, setSortOrder }}>
            {children}
        </FilteredTracksContext.Provider>
    )
}

export const useFilteredTracks = (): FilteredTracksContextType => {
    const context = useContext(FilteredTracksContext)
    if (!context) {
        throw new Error('useFilteredTracks must be used within a FilteredTracksProvider')
    }
    return context
}
