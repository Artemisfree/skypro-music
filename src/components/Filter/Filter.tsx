'use client'

import React, { useState, useEffect, MouseEvent, useCallback } from 'react'
import { getAllTracks } from '../../../api/api'
import styles from './Filter.module.css'
import { Track } from '@/types/types'
import { useFilteredTracks } from '@/contexts/FilteredTracksContext'

interface Position {
	top: number
	left: number
}

const Filter: React.FC = () => {
	const [tracks, setTracks] = useState<Track[]>([])
	const [uniqueAuthors, setUniqueAuthors] = useState<string[]>([])
	const [uniqueGenres, setUniqueGenres] = useState<string[]>([])
	const [activeFilter, setActiveFilter] = useState<string | null>(null)
	const [dropdownPosition, setDropdownPosition] = useState<Position>({
		top: 0,
		left: 0,
	})

	const {
		activeGenres,
		setActiveGenres,
		activeAuthors,
		setActiveAuthors,
		sortOrder,
		setSortOrder,
	} = useFilteredTracks()

	const updateFilters = useCallback(
		(authors?: string[], genres?: string[]) => {
			setUniqueAuthors(authors ?? uniqueAuthors)
			setUniqueGenres(genres ?? uniqueGenres)
		},
		[uniqueAuthors, uniqueGenres]
	)

	useEffect(() => {
		const fetchTracks = async () => {
			try {
				const response = await getAllTracks()
				const data = response.data
				if (Array.isArray(data)) {
					setTracks(data)
					const uniqueValues = extractUniqueValues(data)
					updateFilters(uniqueValues.authors, uniqueValues.genres)
				} else {
					console.error('Неправильный формат данных:', data)
				}
			} catch (error) {
				console.error('Ошибка при получении треков:', error)
			}
		}

		if (tracks.length === 0) {
			fetchTracks()
		}
	}, [tracks.length, updateFilters])

	const extractUniqueValues = (tracks: Track[]) => {
		const authors = new Set<string>()
		const genres = new Set<string>()

		tracks.forEach(track => {
			authors.add(track.author)
			track.genre.forEach(genre => genres.add(genre))
		})

		return {
			authors: Array.from(authors),
			genres: Array.from(genres),
		}
	}

	const handleFilterClick = (
		filter: string,
		event: MouseEvent<HTMLDivElement>
	) => {
		const button = event.currentTarget
		const rect = button.getBoundingClientRect()
		setDropdownPosition({
			top: rect.bottom + window.scrollY + 10,
			left: rect.left + window.scrollX,
		})
		setActiveFilter(prevFilter => {
			if (prevFilter === filter) {
				return null
			} else {
				return filter
			}
		})
	}

	useEffect(() => {
		if (activeFilter !== 'author') setActiveAuthors([])
		if (activeFilter !== 'genre') setActiveGenres([])
		if (activeFilter !== 'year') setSortOrder('default')
	}, [activeFilter])

	const handleGenreClick = (genre: string) => {
		setActiveGenres(prevGenres =>
			prevGenres.includes(genre)
				? prevGenres.filter(g => g !== genre)
				: [...prevGenres, genre]
		)
	}

	const handleAuthorClick = (author: string) => {
		setActiveAuthors(prevAuthors =>
			prevAuthors.includes(author)
				? prevAuthors.filter(a => a !== author)
				: [...prevAuthors, author]
		)
	}

	const handleSortOrderChange = (order: string) => {
		setSortOrder(order)
	}

	return (
		<div className={styles.centerblock__filter}>
			<div className={styles.filter__title}>Искать по:</div>
			<div
				className={`${styles.filter__button} ${
					activeFilter === 'author' ? styles.active : ''
				}`}
				onClick={event => handleFilterClick('author', event)}
			>
				исполнителю
				{activeAuthors.length > 0 && (
					<div className={styles.filter__badge}>{activeAuthors.length}</div>
				)}
			</div>
			{activeFilter === 'author' && (
				<div
					className={`${styles.filter__dropdown} ${styles.dropdown_performer} ${styles.active}`}
					style={{ top: dropdownPosition.top, left: dropdownPosition.left }}
				>
					{uniqueAuthors.map(author => (
						<div
							key={author}
							className={`${styles.filter__item} ${
								activeAuthors.includes(author) ? styles.active : ''
							}`}
							onClick={() => handleAuthorClick(author)}
						>
							{author}
						</div>
					))}
				</div>
			)}
			<div
				className={`${styles.filter__button} ${
					activeFilter === 'year' ? styles.active : ''
				}`}
				onClick={event => handleFilterClick('year', event)}
			>
				дате выпуска
			</div>
			{activeFilter === 'year' && (
				<div
					className={`${styles.filter__dropdown} ${styles.dropdown_year} ${styles.active}`}
					style={{ top: dropdownPosition.top, left: dropdownPosition.left }}
				>
					<div
						className={`${styles.filter__item} ${
							sortOrder === 'default' ? styles.active : ''
						}`}
						onClick={() => handleSortOrderChange('default')}
					>
						По умолчанию
					</div>
					<div
						className={`${styles.filter__item} ${
							sortOrder === 'newest' ? styles.active : ''
						}`}
						onClick={() => handleSortOrderChange('newest')}
					>
						Сначала новые
					</div>
					<div
						className={`${styles.filter__item} ${
							sortOrder === 'oldest' ? styles.active : ''
						}`}
						onClick={() => handleSortOrderChange('oldest')}
					>
						Сначала старые
					</div>
				</div>
			)}
			<div
				className={`${styles.filter__button} ${
					activeFilter === 'genre' ? styles.active : ''
				}`}
				onClick={event => handleFilterClick('genre', event)}
			>
				жанру
			</div>
			{activeFilter === 'genre' && (
				<div
					className={`${styles.filter__dropdown} ${styles.dropdown_genre} ${styles.active}`}
					style={{ top: dropdownPosition.top, left: dropdownPosition.left }}
				>
					{uniqueGenres.map(genre => (
						<div
							key={genre}
							className={`${styles.filter__item} ${
								activeGenres.includes(genre) ? styles.active : ''
							}`}
							onClick={() => handleGenreClick(genre)}
						>
							{genre}
						</div>
					))}
				</div>
			)}
		</div>
	)
}

export default Filter
