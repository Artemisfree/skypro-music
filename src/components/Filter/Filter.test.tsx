import React from 'react'
import renderer from 'react-test-renderer'
import Filter from './Filter'
import { FilteredTracksProvider } from '@/contexts/FilteredTracksContext'



jest.mock('@/app/api', () => ({
	getAllTracks: jest.fn().mockResolvedValue({
		data: [
			{ id: 1, author: 'Author1', genre: ['Genre1'], year: 2020 },
			{ id: 2, author: 'Author2', genre: ['Genre2'], year: 2021 },
		],
	}),
}))

describe('Filter component', () => {
	it('renders correctly and matches snapshot', async () => {
		const component = renderer.create(
			<FilteredTracksProvider>
				<Filter />
			</FilteredTracksProvider>
		)

		await renderer.act(async () => {
			await Promise.resolve()
		})

		let tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})
})
