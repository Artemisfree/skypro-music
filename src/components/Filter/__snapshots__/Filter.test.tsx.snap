// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component renders correctly and matches snapshot 1`] = `
<div
  className="centerblock__filter"
>
  <div
    className="filter__title"
  >
    Искать по:
  </div>
  <div
    className="filter__button "
    onClick={[Function]}
  >
    исполнителю
  </div>
  <div
    className="filter__button "
    onClick={[Function]}
  >
    дате выпуска
  </div>
  <div
    className="filter__button "
    onClick={[Function]}
  >
    жанру
  </div>
</div>
`;
